### PR TITLE
UGDG

### PR DESCRIPTION
**Pull request by Halpfrog**
* Add in dragrams for activity sequence
* Add in manual testing for FindBookCommand
* Add implementations details

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -352,6 +353,33 @@ Each book request is designed to bind to an isbn instead of a book copy. For exa
...
+The following activity diagram summarizes what happens when a user executes a request command:
+
+<img src="images/FindBookCommandActivityDiagram.png" width="650" />
+
+
+#### Design considerations
+
+Each find query can only be one of either the tag, author or title. The feature is designed to return all books that match the predicate created and display them. We chose not to include the isbn as a user searchable query as it is likely that most users would remember the title of the book or the author's name rather than the isbn.
+To account for cases where there might be multiple editions of the same book, the book find will return partial matches. This increases usability as the librarian can find all books that match the title even if it is the first edition or the fifth.
+The feature is also designed to make other features like `borrow` and `request` easier. LibTask can store a large amount of books and users cannot be expected to scroll through them just to find the index of the book they are looking for. `book find` aims to reduce the time spent searching by providing a simple way to search for your book in multiple ways.
```
Comment (aditi2313) [inline docs/DeveloperGuide.md:382]: Well written implementation and design considerations for `book find`

**Review by aditi2313 — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -352,6 +353,33 @@ Each book request is designed to bind to an isbn instead of a book copy. For exa
 
 Associating a book request with multiple book copies introduces some problems. Initially, when a book becomes available, the same reminder message to the user will be printed multiple times, once per book request per book copy. Nevertheless, this is solved by using a `Set` to store reminder messages so that identical reminder messages will not be added multiple times.
 
+### Book Find feature
+
+This feature allows users to search for any books with either the isbn, title or author's name.
```
Comment (yl-ang) [inline docs/DeveloperGuide.md:358]: Minor bug, the find book command doesn't search for ISBN 🤔

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -352,6 +353,33 @@ Each book request is designed to bind to an isbn instead of a book copy. For exa
...
+
+Given below is an example usage scenario and how the request mechanism behaves at each step:
+
+1. The user enters a book find command and provides the parameter for the search query.
+2. `LibTaskParser` creates a new `BookCommandParser` after preliminary processing of user input, which in turns creates a new `FindBookParser`.
+3. `FindBookParser` creates either a `BookAuthorContainsKeywordsPredicatem` or `BookNameContainsKeywordsPredicate` or `BookTagContainsKeywordsPredicate` object `predicate` with the search query.4. 
+4. `FindBookParser` creates a new `FindBookCommand` based on the processed input and passes the predicate on.
+5. `LogicManager` then executes the `FindBookCommand`. 
+6. `FindBookCommand` calls `Model#updateFilteredBookList()` with the `predicate`, resulting in the book list to be updated to display all the books that match the given search query.
+7. Finally, the `FindBookCommand` creates a `CommandResult` and returns it to `LogicManager` to complete the command.
```
Comment (yl-ang) [inline docs/DeveloperGuide.md:371]: Its good to add in the sequence diagram.

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -352,6 +353,33 @@ Each book request is designed to bind to an isbn instead of a book copy. For exa
...
+This feature allows users to search for any books with either the isbn, title or author's name. 
+
+#### Implementation
+The Book find feature is facilitated by the `BookCommandParser`, `FindBookParser` and `FindBookCommand`. 
+
+Given below is an example usage scenario and how the request mechanism behaves at each step:
+
+1. The user enters a book find command and provides the parameter for the search query.
+2. `LibTaskParser` creates a new `BookCommandParser` after preliminary processing of user input, which in turns creates a new `FindBookParser`.
+3. `FindBookParser` creates either a `BookAuthorContainsKeywordsPredicatem` or `BookNameContainsKeywordsPredicate` or `BookTagContainsKeywordsPredicate` object `predicate` with the search query.4.
```
Comment (yl-ang) [inline docs/DeveloperGuide.md:367]: Typo in `BookAuthorContainsKeywordsPredicatem`  and `query.4. `

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -352,6 +353,33 @@ Each book request is designed to bind to an isbn instead of a book copy. For exa
...
+
+#### Implementation
+The Book find feature is facilitated by the `BookCommandParser`, `FindBookParser` and `FindBookCommand`. 
+
+Given below is an example usage scenario and how the request mechanism behaves at each step:
+
+1. The user enters a book find command and provides the parameter for the search query.
+2. `LibTaskParser` creates a new `BookCommandParser` after preliminary processing of user input, which in turns creates a new `FindBookParser`.
+3. `FindBookParser` creates either a `BookAuthorContainsKeywordsPredicatem` or `BookNameContainsKeywordsPredicate` or `BookTagContainsKeywordsPredicate` object `predicate` with the search query.4. 
+4. `FindBookParser` creates a new `FindBookCommand` based on the processed input and passes the predicate on.
```
Comment (yl-ang) [inline docs/DeveloperGuide.md:368]: Should be `predicate`

File: docs/diagrams/FindBookCommandActivityDiagram.puml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+        : Show the book that matches the search queries;
+    else ([else])
+        : Show an empty book list;
+    endif
+else ([else])
+    : Ui shows an error message;
+endif
+stop
+@enduml
+
```
Comment (yl-ang) [inline docs/diagrams/FindBookCommandActivityDiagram.puml:17]: Update as according to the discussion

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (yl-ang) [inline docs/DeveloperGuide.md:367]: Beware of how you going to represent this in the sequence diagram ⚠️

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -352,6 +353,33 @@ Each book request is designed to bind to an isbn instead of a book copy. For exa
...
+7. Finally, the `FindBookCommand` creates a `CommandResult` and returns it to `LogicManager` to complete the command. 
+
+The following activity diagram summarizes what happens when a user executes a request command:
+
+<img src="images/FindBookCommandActivityDiagram.png" width="650" />
+
+
+#### Design considerations
+
+Each find query can only be one of either the tag, author or title. The feature is designed to return all books that match the predicate created and display them. We chose not to include the isbn as a user searchable query as it is likely that most users would remember the title of the book or the author's name rather than the isbn.
```
Comment (yl-ang) [inline docs/DeveloperGuide.md:380]: `The feature is designed to return all books` -> can be phrased to -> `The feature is designed to display all books`

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (yl-ang) [reply docs/DeveloperGuide.md:382]: Well written implementation and design considerations for book find

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1111,8 +1138,30 @@ and borrowed by a patron.
...
+  1. Prerequisites: List all books using the `book list` command. Ensure that there is one book is titled "Harry Potter and the Philosopher's Stone".
+
+  2. Test case: `book find n/Harry Potter and the Philospher's Stone`
+     Expected: One book will be displayed in the book list. 
+
+2. Searching for a book based on tags.
+
+  1. Prerequisites: List all books using the `book list` command. Ensure that one book has the tag "Romance".
+
+  2. Test case: `book find t/Adventure`
```
Comment (yl-ang) [inline docs/DeveloperGuide.md:1156]: Test case: `book find t/Adventure` should be `book find t/Romance`.

**Review by yl-ang — CHANGES_REQUESTED** [review]
Request changes as discussed 💯

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1111,8 +1138,30 @@ and borrowed by a patron.
...
+     Expected: One book will be shown in the book list.
```
Comment (aditi2313) [inline docs/DeveloperGuide.md:1157]: Try being a little more specific. "One book with tag romance will be shown in the book list"

**Review by aditi2313 — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -352,6 +353,41 @@ Each book request is designed to bind to an isbn instead of a book copy. For exa
...
+</div>
+
+The following activity diagram summarizes what happens when a user executes a request command:
+
+<img src="images/FindBookCommandActivityDiagram.png" width="650" />
+
+
+#### Design considerations
+
+Each find query can only be one of either the tag, author or title. The feature is designed to display all books that match the predicate created and display them. We chose not to include the isbn as a user searchable query as it is likely that most users would remember the title of the book or the author's name rather than the isbn.
```
Comment (bingsen0806) [inline docs/DeveloperGuide.md:387]: Good explanation as to why a certain feature is not incldued

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -352,6 +353,41 @@ Each book request is designed to bind to an isbn instead of a book copy. For exa
...
+To account for cases where there might be multiple editions of the same book, the book find will return partial matches. This increases usability as the librarian can find all books that match the title even if it is the first edition or the fifth.
```
Comment (bingsen0806) [inline docs/DeveloperGuide.md:388]: Good explanation has to why partial matches are used instead of full matches

**Review by bingsen0806 — APPROVED** [review]
LGTM!

**Pull request merged**